pkg/activator/net/throttler.go

```diff
@@ -114,67 +114,52 @@ func newRevisionThrottler(revID types.NamespacedName,
 	}
 }
 
-// pickP2C implements power of two choices algorithm for fast load balancing.
-// See here: https://www.eecs.harvard.edu/~michaelm/postscripts/mythesis.pdf.
-// The function picks a target to send request to, given list of `podIPTracker` objects, a
-// callback to release the slot is returned as well.
-func pickP2C(tgts []*podIPTracker) (string, func()) {
-	var t1, t2 *podIPTracker
-	switch len(tgts) {
-	case 0:
-		return "", nil
-	case 1:
-		t1, t2 = tgts[0], tgts[0]
-	case 2:
-		t1, t2 = tgts[0], tgts[1]
-	default:
-		i1 := rand.Intn(len(tgts))
-		t1 = tgts[i1]
-		i2 := rand.Intn(len(tgts))
-		for i1 == i2 {
-			i2 = rand.Intn(len(tgts))
-		}
-		t2 = tgts[i2]
-	}
-	// Note that this is not guaranteed to be precise, due to the case
-	// that Load here and Add below are not atomic.
-	if atomic.LoadInt32(&t1.requests) > atomic.LoadInt32(&t2.requests) {
-		t1 = t2
+// pickPod picks the first tracker that has open capacity if container concurrency
+// if limited, random pod otherwise.
+// TODO(vagabaov): make this locally ideal, rather than mostly ideal.
+func pickPod(tgs []*podIPTracker, cc int64) *podIPTracker {
+	// Infinite capacity, pick random. We have to do this
+	// otherwise _all_ the requests will go to the first pod
+	// since it has unlimited capacity.
+	if cc == 0 {
+		return tgs[rand.Intn(len(tgs))]
 	}
-	atomic.AddInt32(&t1.requests, 1)
-	return t1.dest, func() {
-		atomic.AddInt32(&t1.requests, -1)
+	for _, t := range tgs {
+		if t.b.HasCapacity() {
+			return t
+		}
 	}
+	return nil
 }
 
-func noop() {}
-
-// Returns a dest after incrementing its request count and a completion callback
-// to be called after request completion. If no dest is found it returns "", nil.
-func (rt *revisionThrottler) acquireDest() (string, func()) {
+// Returns a dest that at the moment of choosing had an open slot
+// for request.
+func (rt *revisionThrottler) acquireDest() *podIPTracker {
 	rt.mux.RLock()
 	defer rt.mux.RUnlock()
 
-	// This is intended to be called only after performing a read lock check on clusterIPDest
-	if rt.clusterIPDest != "" {
-		return rt.clusterIPDest, noop
+	if rt.clusterIPDest != nil {
+		return rt.clusterIPDest
 	}
-	return pickP2C(rt.assignedTrackers)
+	return pickPod(rt.assignedTrackers, rt.containerConcurrency)
 }
 
 func (rt *revisionThrottler) try(ctx context.Context, function func(string) error) error {
 	var ret error
 
 	if err := rt.breaker.Maybe(ctx, func() {
```

Comment:
Do we need/want the outer breaker at this point? I see why it's there and we have a growing codebase here. I do wonder though if we can hide it as an implementation detail and have `acquireDest` return a `podIPTracker` that represents the serviceIP itself here.

Reply:
I don't think so. At least no as written. Since if we're out of capacity we will need to pick a victim pod breaker to queue on it, which might be a wrong pod to take (i.e. it's full and all of its requests are long running).

Reply:
Good point, agreed.